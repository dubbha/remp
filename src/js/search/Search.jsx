import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import SearchHeader from '../common/searchHeader';
import Result from '../common/result';
import List from '../common/list';
import Footer from '../common/footer';
import filmPropShape from '../common/utils/propShapes';
import * as actions from '../common/store/actions';
import * as selectors from '../common/store/selectors';

export class Search extends Component {
  static propTypes = {
    match: PropTypes.shape({
      params: PropTypes.shape({
        query: PropTypes.string,
      }),
    }).isRequired,
    history: PropTypes.shape({
      push: PropTypes.func,
    }).isRequired,
    query: PropTypes.string.isRequired,
    results: PropTypes.arrayOf(PropTypes.shape(filmPropShape)).isRequired,
    searchBy: PropTypes.string.isRequired,
    sortBy: PropTypes.string.isRequired,
    searchByParams: PropTypes.arrayOf(PropTypes.string).isRequired,
    sortByParams: PropTypes.arrayOf(PropTypes.string).isRequired,
    isLoading: PropTypes.bool.isRequired,
    searchByTitle: PropTypes.func.isRequired,
    searchByDirector: PropTypes.func.isRequired,
    setQuery: PropTypes.func.isRequired,
    setResults: PropTypes.func.isRequired,
    clearResults: PropTypes.func.isRequired,
    setSearchBy: PropTypes.func.isRequired,
    setSortBy: PropTypes.func.isRequired,
  };

  static fetchData = (dispatch, match) => {
    if (match.params.query) {
      const paramsQuery = decodeURIComponent(match.params.query);
      dispatch(actions.setQuery(paramsQuery));
      return dispatch(actions.searchByDirector(paramsQuery));
    }
    return Promise.resolve();
  }

  componentWillMount() {
    const {
      match: { params },
      query,
      searchBy,
      sortBy,
      searchByTitle,
      searchByDirector,
      setQuery,
    } = this.props;

    if (params.query) {
      const paramsQuery = decodeURIComponent(params.query);
      if (paramsQuery !== query) {
        setQuery(paramsQuery);
        if (searchBy === 'director') {
          searchByDirector(paramsQuery, sortBy);
        } else {
          searchByTitle(paramsQuery, sortBy);
        }
      }
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.match.params.query !== this.props.match.params.query) {
      const { searchBy, sortBy, setQuery, clearResults, searchByDirector, searchByTitle } = nextProps;

      if (nextProps.match.params.query) {
        const nextPropsQuery = decodeURIComponent(nextProps.match.params.query);

        setQuery(nextPropsQuery);
        if (searchBy === 'director') {
          searchByDirector(nextPropsQuery, sortBy);
        } else {
          searchByTitle(nextPropsQuery, sortBy);
        }
      } else {
        setQuery('');
        clearResults();
      }
    }
  }

  handleSearch = (e) => {
    const {
      history,
      query,
      searchBy,
      sortBy,
      searchByTitle,
      searchByDirector,
      clearResults,
    } = this.props;

    e.preventDefault(); // submitting the form to support search on enter

    if (query) {
      history.push(`/search/${encodeURIComponent(query)}`);
      if (searchBy === 'director') {
        searchByDirector(query, sortBy);
      } else {
        searchByTitle(query, sortBy);
      }
    } else {
      history.push('/search');
      clearResults();
    }
  }

  handleQueryChange = (e) => {
    this.props.setQuery(e.target.value);
  }

  handleSelectFilm = (film) => {
    const { history } = this.props;

    window.scrollTo(0, 0);
    history.push(`/film/${film.id}/${encodeURIComponent(film.title)}`);
  }

  handleSearchByChange = (searchBy) => {
    this.props.setSearchBy(searchBy);
  }

  handleSortByChange = (sortBy) => {
    const { setSortBy, results, setResults } = this.props;

    setSortBy(sortBy);
    setResults(results, sortBy);
  }

  render() {
    const {
      query,
      results,
      searchBy,
      sortBy,
      searchByParams,
      sortByParams,
      isLoading,
    } = this.props;

    return (
      <div className="app__container">
        <SearchHeader
          query={query}
          onQueryChange={this.handleQueryChange}
          onSearch={this.handleSearch}
          searchBy={searchBy}
          onSearchByChange={this.handleSearchByChange}
          searchByParams={searchByParams}
        />
        <Result
          results={results}
          sortBy={sortBy}
          onSortByChange={this.handleSortByChange}
          sortByParams={sortByParams}
        />
        <List
          results={results}
          onSelectFilm={this.handleSelectFilm}
          isLoading={isLoading}
        />
        <Footer />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  query: selectors.querySelector(state),
  results: selectors.resultsSelector(state),
  searchBy: selectors.searchBySelector(state),
  sortBy: selectors.sortBySelector(state),
  searchByParams: selectors.searchByParamsSelector(state),
  sortByParams: selectors.sortByParamsSelector(state),
  isLoading: selectors.isLoadingSelector(state),
});

const mapDispatchToProps = {
  searchByTitle: actions.searchByTitle,
  searchByDirector: actions.searchByDirector,
  setQuery: actions.setQuery,
  setResults: actions.setResults,
  clearResults: actions.clearResults,
  setSearchBy: actions.setSearchBy,
  setSortBy: actions.setSortBy,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Search);
