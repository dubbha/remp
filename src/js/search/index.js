import reducer from '../common/store/reducer';
import * as actions from '../common/store/actions';
import * as selectors from '../common/store/selectors';

export { default } from './Search';

export {
  reducer,
  actions,
  selectors,
};
