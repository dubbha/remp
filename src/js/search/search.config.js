export const searchByParams = ['title', 'director'];
export const sortByParams = ['release date', 'rating'];

export const defaultSearchBy = searchByParams[0];
export const defaultSortBy = sortByParams[0];

const sortByApiMap = {
  'release date': 'release_year',
  rating: 'rating',
};

export const sortFnFabric = sortBy => (a, b) => b[sortByApiMap[sortBy]] - a[sortByApiMap[sortBy]];
