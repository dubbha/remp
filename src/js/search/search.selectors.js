import { createSelector } from 'reselect';

const searchStateSelector = state => state.search;

export const querySelector = createSelector(
  searchStateSelector,
  ({ query }) => query,
);

export const resultsSelector = createSelector(
  searchStateSelector,
  ({ results }) => results,
);

export const searchBySelector = createSelector(
  searchStateSelector,
  ({ searchBy }) => searchBy,
);

export const sortBySelector = createSelector(
  searchStateSelector,
  ({ sortBy }) => sortBy,
);

export const searchByParamsSelector = createSelector(
  searchStateSelector,
  ({ searchByParams }) => searchByParams,
);

export const sortByParamsSelector = createSelector(
  searchStateSelector,
  ({ sortByParams }) => sortByParams,
);

export const isLoadingSelector = createSelector(
  searchStateSelector,
  ({ isLoading }) => isLoading,
);
