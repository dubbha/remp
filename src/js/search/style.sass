.search
    background-image: url('../../img/bg.jpg')
    background-size: cover
    min-height: 450px
    padding: 20px 50px

    &__header
        margin: 50px 0 20px 0
        text-transform: uppercase
        color: #fff
        font-family: Helvetica
        font-weight: bold

    &__input
        background-color: #000
        border-top: none
        border-right: none
        border-bottom: 1px solid #f55263
        border-left: none
        color: #fff
        width: 100%
        padding: 10px

        &:focus
            outline: none
