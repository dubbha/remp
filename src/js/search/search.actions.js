import axios from 'axios';
import { defaultSearchBy, defaultSortBy, sortFnFabric } from './search.config';

export const actionTypes = {
  SET_QUERY: 'SET_QUERY',
  SET_RESULTS: 'SET_RESULTS',
  CLEAR_RESULTS: 'CLEAR_RESULTS',
  SET_SEARCH_BY: 'SET_SEARCH_BY',
  SET_SORT_BY: 'SET_SORT_BY',
  SET_IS_LOADING: 'SET_IS_LOADING',
};

export const setQuery = query => ({
  type: actionTypes.SET_QUERY,
  query,
});

export const setResults = (results, sortBy) => ({
  type: actionTypes.SET_RESULTS,
  results: Array.isArray(results) ? results.sort(sortFnFabric(sortBy)) : [results],
});

export const clearResults = () => ({
  type: actionTypes.CLEAR_RESULTS,
});

export const setSearchBy = searchBy => ({
  type: actionTypes.SET_SEARCH_BY,
  searchBy,
});

export const setSortBy = sortBy => ({
  type: actionTypes.SET_SORT_BY,
  sortBy,
});

export const setIsLoading = isLoading => ({
  type: actionTypes.SET_IS_LOADING,
  isLoading,
});

export const search = (query, searchBy = defaultSearchBy, sortBy = defaultSortBy) => (dispatch) => {
  dispatch(setIsLoading(true));
  return axios.get('/api', { params: { [searchBy]: query } })
    .then((res) => {
      dispatch(setResults(res.data, sortBy));
      dispatch(setIsLoading(false));
    })
    .catch(() => {
      dispatch(clearResults());
      dispatch(setIsLoading(false));
    });
};
