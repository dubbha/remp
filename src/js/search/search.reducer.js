import { actionTypes } from './search.actions';
import { searchByParams, sortByParams, defaultSearchBy, defaultSortBy } from './search.config';

const initialState = {
  query: '',
  results: [],
  searchByParams,
  sortByParams,
  searchBy: defaultSearchBy,
  sortBy: defaultSortBy,
  isLoading: false,
};

const searchReducer = (state = initialState, action) => {
  switch (action.type) {
    case actionTypes.SET_QUERY:
      return {
        ...state,
        query: action.query,
      };
    case actionTypes.SET_RESULTS:
      return {
        ...state,
        results: action.results,
      };
    case actionTypes.CLEAR_RESULTS:
      return {
        ...state,
        results: [],
      };
    case actionTypes.SET_SEARCH_BY:
      return {
        ...state,
        searchBy: action.searchBy,
      };
    case actionTypes.SET_SORT_BY:
      return {
        ...state,
        sortBy: action.sortBy,
      };
    case actionTypes.SET_IS_LOADING:
      return {
        ...state,
        isLoading: action.isLoading,
      };
    default:
      return state;
  }
};

export default searchReducer;

