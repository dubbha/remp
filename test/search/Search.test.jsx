import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { connect } from 'react-redux';
import { Search } from 'search/Search';

jest.mock('react-redux', () => ({
  connect: jest.fn(() => jest.fn()),
}));

jest.mock('common/store/selectors', () => ({
  querySelector: jest.fn(state => state.query),
  resultsSelector: jest.fn(state => state.results),
  searchBySelector: jest.fn(state => state.searchBy),
  sortBySelector: jest.fn(state => state.sortBy),
  searchByParamsSelector: jest.fn(state => state.searchByParams),
  sortByParamsSelector: jest.fn(state => state.sortByParams),
  isLoadingSelector: jest.fn(state => state.isLoading),
}));

jest.mock('common/store/actions', () => ({
  setQuery: jest.fn(query => `setQuery action for ${query}`),
  searchByDirector: jest.fn(query => `searchByDirector action for ${query}`),
}));

describe('Search', () => {
  describe('react-redux connector', () => {
    it('should map state to props', () => {
      const mapStateToProps = connect.mock.calls[0][0];

      const state = {
        query: 'QUERY',
        results: [
          { id: 1, title: 'TITLE1' },
          { id: 2, title: 'TITLE2' },
          { id: 3, title: 'TITLE3' },
        ],
        searchBy: 'director',
        sortBy: 'rating',
        searchByParams: ['title', 'director'],
        sortByParams: ['release date', 'rating'],
        isLoading: false,
      };

      expect(mapStateToProps(state)).toEqual({
        query: 'QUERY',
        results: [
          { id: 1, title: 'TITLE1' },
          { id: 2, title: 'TITLE2' },
          { id: 3, title: 'TITLE3' },
        ],
        searchBy: 'director',
        sortBy: 'rating',
        searchByParams: ['title', 'director'],
        sortByParams: ['release date', 'rating'],
        isLoading: false,
      });
    });
  });

  describe('unconnected component', () => {
    const event = {
      preventDefault: jest.fn(),
      target: {
        value: 'VALUE',
      },
    };

    let props;
    beforeEach(() => {
      props = {
        match: { params: { query: 'QUERY' } },
        history: { push: jest.fn() },
        query: 'QUERY',
        results: [
          { id: 1, title: 'TITLE1' },
          { id: 2, title: 'TITLE2' },
          { id: 3, title: 'TITLE3' },
        ],
        searchBy: 'director',
        sortBy: 'rating',
        searchByParams: ['title', 'director'],
        sortByParams: ['release date', 'rating'],
        searchByTitle: jest.fn(),
        searchByDirector: jest.fn(),
        setQuery: jest.fn(),
        setResults: jest.fn(),
        clearResults: jest.fn(),
        setSearchBy: jest.fn(),
        setSortBy: jest.fn(),
        isLoading: false,
      };
    });

    it('should render successfully', () => {
      const wrapper = shallow(<Search {...props} />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should set query if params query differs from stored query on mount', () => {
      props.query = '';
      const instance = new Search(props);

      instance.componentWillMount();

      expect(props.setQuery).toBeCalledWith('QUERY');
    });

    it('should search by director if params query differs from stored query on mount', () => {
      props.query = '';
      const instance = new Search(props);

      instance.componentWillMount();

      expect(props.searchByDirector).toBeCalledWith('QUERY', 'rating');
    });

    it('should search by title if params query differs from stored query on mount', () => {
      props.query = '';
      props.searchBy = 'title';
      const instance = new Search(props);

      instance.componentWillMount();

      expect(props.searchByTitle).toBeCalledWith('QUERY', 'rating');
    });

    it('should not set query if params query is empty on mount', () => {
      props.match.params.query = '';
      const instance = new Search(props);

      instance.componentWillMount();

      expect(props.setQuery).not.toBeCalled();
    });

    it('should set query if route query param changed on props change', () => {
      const instance = new Search(props);

      instance.componentWillReceiveProps({
        match: { params: { query: 'NEW_QUERY' } },
        searchBy: props.searchBy,
        sortBy: props.sortBy,
        setQuery: props.setQuery,
        clearResults: props.clearResults,
        searchByDirector: props.searchByDirector,
        searchByTitle: props.searchByTitle,
      });

      expect(props.setQuery).toBeCalledWith('NEW_QUERY');
    });

    it('should search by director if route query param changed on props change', () => {
      const instance = new Search(props);

      instance.componentWillReceiveProps({
        match: { params: { query: 'NEW_QUERY' } },
        searchBy: props.searchBy,
        sortBy: props.sortBy,
        setQuery: props.setQuery,
        clearResults: props.clearResults,
        searchByDirector: props.searchByDirector,
        searchByTitle: props.searchByTitle,
      });

      expect(props.searchByDirector).toBeCalledWith('NEW_QUERY', 'rating');
    });

    it('should search by title if route query param changed on props change', () => {
      const instance = new Search(props);

      instance.componentWillReceiveProps({
        match: { params: { query: 'NEW_QUERY' } },
        searchBy: 'title',
        sortBy: props.sortBy,
        setQuery: props.setQuery,
        clearResults: props.clearResults,
        searchByDirector: props.searchByDirector,
        searchByTitle: props.searchByTitle,
      });

      expect(props.searchByTitle).toBeCalledWith('NEW_QUERY', 'rating');
    });

    it('should set query if query changed to empty string on props change', () => {
      const instance = new Search(props);

      instance.componentWillReceiveProps({
        match: { params: { query: '' } },
        searchBy: props.searchBy,
        sortBy: props.sortBy,
        setQuery: props.setQuery,
        clearResults: props.clearResults,
        searchByDirector: props.searchByDirector,
        searchByTitle: props.searchByTitle,
      });

      expect(props.setQuery).toBeCalledWith('');
    });

    it('should clear results if query changed to empty string on props change', () => {
      const instance = new Search(props);

      instance.componentWillReceiveProps({
        match: { params: { query: '' } },
        searchBy: props.searchBy,
        sortBy: props.sortBy,
        setQuery: props.setQuery,
        clearResults: props.clearResults,
        searchByDirector: props.searchByDirector,
        searchByTitle: props.searchByTitle,
      });

      expect(props.clearResults).toBeCalled();
    });

    it('should not set query if query has not changed on props change', () => {
      const instance = new Search(props);

      instance.componentWillReceiveProps({
        match: { params: { query: 'QUERY' } },
        searchBy: props.searchBy,
        sortBy: props.sortBy,
        setQuery: props.setQuery,
        clearResults: props.clearResults,
        searchByDirector: props.searchByDirector,
        searchByTitle: props.searchByTitle,
      });

      expect(props.setQuery).not.toBeCalled();
    });

    it('should push to history on handle search', () => {
      const wrapper = shallow(<Search {...props} />);

      wrapper.find('SearchHeader').simulate('search', event);

      expect(props.history.push).toBeCalledWith('/search/QUERY');
    });

    it('should search by director on handle search', () => {
      const wrapper = shallow(<Search {...props} />);

      wrapper.find('SearchHeader').simulate('search', event);

      expect(props.searchByDirector).toBeCalledWith('QUERY', props.sortBy);
    });

    it('should search by title on handle search', () => {
      props.searchBy = 'title';
      const wrapper = shallow(<Search {...props} />);

      wrapper.find('SearchHeader').simulate('search', event);

      expect(props.searchByTitle).toBeCalledWith('QUERY', props.sortBy);
    });

    it('should push to history if query is empty on handle search', () => {
      props.query = '';
      const wrapper = shallow(<Search {...props} />);

      wrapper.find('SearchHeader').simulate('search', event);

      expect(props.history.push).toBeCalledWith('/search');
    });

    it('should clear results if query is empty on handle search', () => {
      props.query = '';
      const wrapper = shallow(<Search {...props} />);

      wrapper.find('SearchHeader').simulate('search', event);

      expect(props.clearResults).toBeCalledWith();
    });

    it('should handle query change', () => {
      const wrapper = shallow(<Search {...props} />);

      wrapper.find('SearchHeader').simulate('queryChange', event);

      expect(props.setQuery).toBeCalledWith('VALUE');
    });

    it('should scroll to top on select film', () => {
      jest.spyOn(window, 'scrollTo');

      const wrapper = shallow(<Search {...props} />);

      wrapper.find('List').simulate('selectFilm', { id: 1, title: 'TITLE' });

      expect(global.window.scrollTo).toBeCalledWith(0, 0);
    });

    it('should push to history select film', () => {
      const wrapper = shallow(<Search {...props} />);

      wrapper.find('List').simulate('selectFilm', { id: 1, title: 'TITLE' });

      expect(props.history.push).toBeCalledWith('/film/1/TITLE');
    });

    it('should handle searchBy change', () => {
      const wrapper = shallow(<Search {...props} />);

      wrapper.find('SearchHeader').simulate('searchByChange', 'NEW_SEARCH_BY');

      expect(props.setSearchBy).toBeCalledWith('NEW_SEARCH_BY');
    });

    it('should handle sortBy change', () => {
      const wrapper = shallow(<Search {...props} />);

      wrapper.find('Result').simulate('sortByChange', 'NEW_SORT_BY');

      expect(props.setSortBy).toBeCalledWith('NEW_SORT_BY');
    });

    it('should sort results on handle sortBy change', () => {
      const wrapper = shallow(<Search {...props} />);

      wrapper.find('Result').simulate('sortByChange', 'NEW_SORT_BY');

      expect(props.setResults).toBeCalledWith(props.results, 'NEW_SORT_BY');
    });
  });

  describe('fetchData static method', () => {
    let match;
    let dispatch;

    beforeEach(() => {
      match = { params: { query: 'QUERY' } };
      dispatch = jest.fn();
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    it('should set query on server-side rendering', () => {
      Search.fetchData(dispatch, match);

      expect(dispatch).toBeCalledWith('setQuery action for QUERY');
    });

    it('should search by director on server-side rendering', () => {
      Search.fetchData(dispatch, match);

      expect(dispatch).toBeCalledWith('searchByDirector action for QUERY');
    });

    it('should not dispatch if query is empty on server-side rendering', () => {
      match.params.query = '';
      Search.fetchData(dispatch, match);

      expect(dispatch).not.toBeCalled();
    });
  });
});
