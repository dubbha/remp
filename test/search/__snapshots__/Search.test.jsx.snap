// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search unconnected component should render successfully 1`] = `
<div
  className="app__container"
>
  <SearchHeader
    onQueryChange={[Function]}
    onSearch={[Function]}
    onSearchByChange={[Function]}
    query="QUERY"
    searchBy="director"
    searchByParams={
      Array [
        "title",
        "director",
      ]
    }
  />
  <Result
    film={null}
    onSortByChange={[Function]}
    results={
      Array [
        Object {
          "id": 1,
          "title": "TITLE1",
        },
        Object {
          "id": 2,
          "title": "TITLE2",
        },
        Object {
          "id": 3,
          "title": "TITLE3",
        },
      ]
    }
    sortBy="rating"
    sortByParams={
      Array [
        "release date",
        "rating",
      ]
    }
  />
  <List
    film={Object {}}
    isLoading={false}
    onSelectFilm={[Function]}
    results={
      Array [
        Object {
          "id": 1,
          "title": "TITLE1",
        },
        Object {
          "id": 2,
          "title": "TITLE2",
        },
        Object {
          "id": 3,
          "title": "TITLE3",
        },
      ]
    }
  />
  <Footer />
</div>
`;
